src/partition/repart.c

```diff
@@ -2362,11 +2536,23 @@ static int context_dump(Context *context, const char *node) {
         if (arg_pretty == 0 && FLAGS_SET(arg_json_format_flags, JSON_FORMAT_OFF))
                 return 0;
 
+        /* If we're outputting JSON, only dump after doing all operations so we can include the roothashes
+         * in the output.  */
+        if (!late && !FLAGS_SET(arg_json_format_flags, JSON_FORMAT_OFF))
+                return 0;
+
+        /* If we're not outputting JSON, only dump again after doing all operations if there are any
+         * roothashes that we need to communicate to the user. */
+        if (late && FLAGS_SET(arg_json_format_flags, JSON_FORMAT_OFF) && !context_has_roothash(context))
+                return 0;
+
         r = context_dump_partitions(context, node);
         if (r < 0)
                 return r;
 
-        if (arg_json_format_flags & JSON_FORMAT_OFF) {
+        /* Make sure we only write the partition bar once, even if we're writing the partition table twice to
+         * communicate roothashes. */
+        if (arg_json_format_flags & JSON_FORMAT_OFF && !late) {
```

Comment:
if you mix logical and binary &/&& in one expression, please use (), it's easier on the eye even if not needed. FLAGS_SET() also works